### ISSUE TITLE
Datatype triple patterns

### ISSUE THREAD

**Issue by IS4Code** [w3c/sparql-12#182]
## Why?

Presently, SPARQL triple syntax does not offer enough granularity when matching literals ‒ the object of a property can be specified in a triple pattern as (among others) a variable or a fixed literal, but it cannot be something in between. If you want to find triples based on the datatypes, you have to match all of them, and then filter:
```sparql
?s ?p ?o .
FILTER (DATATYPE(?o) = xsd:integer)
```
A trivial SPARQL engine implementation would load all triples in existence into a set and then filter them based on the criteria. 
It is possible, of course, to optimize this query and retrieve all matching triples as a single step, but not all engines might do that, and still this is something that could be expressed more concisely.

## Proposed solution

I propose an extension of the node syntax like this:

```sparql
?s ?p ?o^^xsd:integer .
```
Such a query would only match triples with the `xsd:integer` datatype, and bind `?o` to the lexical value of the literal (as a simple literal/`xsd:string`).

Of course there are other options for the object:
```sparql
# any literal with this lexical value, regardless of datatype (?dt is bound to IRI)
?s ?p "10"^^?dt .

# any literal (?o is simple literal, ?dt is IRI)
?s ?p ?o^^?dt .
```

I think this would add more expressiveness to the language, and a possibility for better optimizations to triple store that can index by the datatype and evaluate such queries efficiently.

#### Language-tagged literals

These are literals as well, but I am unsure whether they should be matched by a pattern like `?o^^?dt`. However, if they are allowed, I think the natural solution (if not deemed too convoluted) might be to bind `?dt` to a literal with the `xsd:language` datatype, essentially treating something like `"hello world"@en` as "`"hello world"^^"en"^^xsd:language`" (*not* a proposed syntax), binding `?o` to `"hello world"` and `?dt` to `"en"^^xsd:language`. After all, a language tag *is* the datatype of a language-tagged literal, at least syntactically.

Other SPARQL functions, such as `STRDT`, could be modified to allow either `IRI` or `xsd:language` as the datatype.

## Considerations for backward compatibility

I don't think `^^` in this position could have been valid previously, so all existing valid queries should remain valid and unambiguous.

**Comment by namedgraph** [comment]
Isn't [`rdfs:langString`](https://www.w3.org/TR/rdf-schema/#ch_langstring) the datatype of language-typed literals?

**Comment by IS4Code** [comment]
> Isn't [`rdfs:langString`](https://www.w3.org/TR/rdf-schema/#ch_langstring) the datatype of language-typed literals?

Semantically yes, but it is not written in the triple. Such a thing should be produced only from something like `"hello world"^^rdf:langString`, not from an actual language-tagged literal.

Specifically, I was aiming for `?o^^?dt` to produce something where `STRDT(?o, ?dt)` gives back the original object. With `rdf:langString`, information about the actual language tag is lost.

**Comment by redmer** [comment]
This proposal would make working with typed strings better, so I appreciate it. Combined with `?o@?lang` to match the language tag, it would match both ways that (typed and language tagged) strings are available in SPARQL and Turtle. Then you'd use `STRDT()` to reconstitute a string from `?o^^?dt` and `STRLANG()` with `?o@?lang`. This was proposed in #17.

Having `STRDT(?o, "en^^xsd:language)` return `"hello world"@en` would be a (breaking?) change from SPARQL 1.1. 

Keeping `datatype(""@en) = ?dt = rdfs:langString` the same would give fewer differences with the [Turtle parsing rules](https://www.w3.org/TR/turtle/#handle-RDFLiteral) and SPARQL's [`datatype()`](https://www.w3.org/TR/sparql11-query/#func-datatype). It would indeed entail that `^^?dt` "shadows" the language tag information. But with also a parallel literal syntax to match languages, this may not be a true problem.

Looking at other proposals, #34 suggest `?o@* to match all language tagged (≟ only `rdfs:langString` datatype) strings or `?o^^*` for all datatype tagged (≟ all non-`xsd:string` and/or all non-`rdfs:langString`). #112 looks only tangentially related. 

I think that would result in, for the following data, the bindings as in the following table:

```turtle
:s :p "Hello, world!"@en .
```

| SPARQL                     | ?o                   | ?dt               | ?lang  |
| --------------------------- | -------------------- | ----------------- | ------ |
| `:s :p ?o`                  | `"Hello, world!"@en` |
| `:s :p ?o^^?dt`             | `"Hello, world!"`    | `rdfs:langString` |
| `:s :p ?o^^rdfs:langString` | `"Hello, world!"`    |
| `:s :p ?o@?lang`            | `"Hello, world!"`    |                   | `"en"` |
| `:s :p ?o@en`               | `"Hello, world!"`    |
| ? `:s :p ?o^^?dt , ?o@?lang`   | `"Hello, world!"`     | `rdfs:langString` | `"en"`   |

**Comment by VladimirAlexiev** [comment]
@IS4Code @redmer 
Do you actually have use cases for all of this?
What is the use of "give me all integers" or "give me all en-US strings"?

And here are some more considerations for you to think about:
- If the lang is `en-US` should it be returned as `en-US` (proper case as per [RFC 5646](https://datatracker.ietf.org/doc/html/rfc5646#section-2.1.1) or `en-us` (lowercase)? There's a reason `langMatches()` exists
- If the lang is `en-Latn` should it be returned as `en` because script `Latn` is default for that lang (see [RFC 5646 3.1.9 Suppress-Script Field](https://datatracker.ietf.org/doc/html/rfc5646#section-3.1.9))
- I have a useful case that would require DISJUNCTION: "give me the plain strings or English strings"
- XSD datatypes have a hierarchy (eg `xsd:int<xsd:integer<xsd:decimal`), should it be taken into account? Why or why not?
- the last example `:s :p ?o^^?dt , ?o@?lang` is bogus, 
imagine that you have several values with the same string value `"Hello, world!"@en, "Hello, world!"@de, "Hello, world!"` 
and you'll see a Cartesian Explosion forming

**Comment by IS4Code** [comment]
Statistics are what comes to my mind immediately ‒ if you have a large multi-language dataset, you could check immediately if there is, say, a deficiency in a particular language compared to another one, e.g.
```sparql
SELECT (COUNT(?o) AS ?c)
WHERE {
  ?s ?p ?o@de-DE .
}
```
Being able to partition by a language directly could be more desirable than filtering all objects.

Similarly, you may want to retrieve all `rdf:XMLLiteral` literals to check for a particular conformance, and so on.

To address some of the points:
* I think the form in which the language tag should be treated here is already determined by the existing operations ‒ `?o@de-DE` should match all nodes that would be matched by some <code>"<i>text</i>"@de-DE</code>, with identical normalization. Similarly, if `?o^^?dt` retrieves a language tag, it should have the same content as if retrieved using `LANG`. I can see someone wanting to match `en-Latn` just with `@en`, but that is something that already concerns normal language-tagged literals and not just this extension.
* Disjunction is useful and should be doable with `UNION`.
* Regarding datatype hierarchies, the answer is basically the same as for language tags ‒ `?o^^xsd:decimal` should match anything matched by some <code>"<i>value</i>"^^xsd:decimal</code>. Some implementations do match `"5"^^xsd:integer` with `"5"^^xsd:decimal`, some don't and this should be again consistent with how existing patterns behave. Of course, that would mean that
  ```sparql
  ?s ?p ?o .
  FILTER (DATATYPE(?o) = xsd:decimal)
  ```
  is not strictly equivalent to `?s ?p ?o^^xsd:decimal` ‒ the first pattern only selects triples with the exact datatype, while the new one *could* also select triples with `xsd:integer`, `xsd:int` and so on depending on how the implementation compares the datatypes when matching literals. However, I think that is still desirable, since expressing that sort of behaviour with existing options is a bit shaky (`FILTER (?o = STRDT(STR(?o), xsd:decimal)`) and becomes an even less viable target of optimizations.

**Comment by VladimirAlexiev** [comment]
> deficiency in a particular language

Don't you also need to compare to the prevalence of other languages?
BTW, a good way to express such statistics about a dataset is http://ldf.fi/void-ext ("object language partition" and the like).

> rdf:XMLLiteral literals to check for a particular conformance,

Agree. 
We have related issues in the Electrical CIM:
- https://github.com/Sveino/Inst4CIM-KG/issues/21
- https://github.com/Sveino/Inst4CIM-KG/issues/153
- https://github.com/Sveino/Inst4CIM-KG/labels/typography

But my practical experience is that you find all kinds of markup in descriptions (eg HTML tags or HTML-encoded chars), even though the literal is not marked as `rdf:XMLLiteral` or `rdf:HTMLLiteral`.

**Comment by IS4Code** [comment]
Yes, I did not include other languages in the query for the sake of simplicity (for example you could count all `en` texts and express any other desired language as a ratio of that).
